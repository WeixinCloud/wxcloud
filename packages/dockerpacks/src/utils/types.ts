export type NonEmptyArray<T> = [T, ...T[]];

export function isGenericNonEmptyArray(input: any): input is NonEmptyArray<unknown> {
  return Array.isArray(input) && isNonEmptyArray(input);
}

export function isNonEmptyArray<T = unknown>(input: Array<T>): input is NonEmptyArray<T> {
  return input.length > 0;
}
