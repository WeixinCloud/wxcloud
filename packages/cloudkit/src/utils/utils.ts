/**
 * isMatchMajorVersion('^1.2.3', '1') -> true
 *
 * support ~version & ^version
 * @param version currentVersion
 * @param majorVersion target majorVersion
 */
export const isMatchMajorVersion = (version: string, majorVersion: string) => {
  return (
    version.startsWith(`^${majorVersion}`) ||
    version.startsWith(`~${majorVersion}`) ||
    version.startsWith(majorVersion)
  );
};

/**
 * safely get deps value from packageJson
 * @param packageJson
 * @param key
 * @returns
 */
export const safeGetDepsFromPkgJSON = (packageJson: any, key: string) => {
  return packageJson?.dependencies?.[key] || packageJson?.devDependencies?.[key];
};

export const crossPlatformNpm = /^win/.test(process.platform) ? 'npm.cmd' : 'npm';
