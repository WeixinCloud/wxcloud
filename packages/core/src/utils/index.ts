export { logger } from './logger';
export { safeRequire } from './safeRequire';
export { preprocessBaseConfig } from './baseconfig';
