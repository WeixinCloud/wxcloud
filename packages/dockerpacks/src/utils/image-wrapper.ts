import { ImageTag, MatchedImage } from '@api/image';

export class ImageWrapper {
  constructor(private readonly image: MatchedImage) {}

  getFullVersion() {
    return this.image.fullVersion;
  }

  getFullVersionTag() {
    return this.image.fullVersionTag;
  }

  getMostGeneralTag() {
    if (!this.image.matchedTags.length) {
      throw new Error('unexpected matchedTags of image');
    }
    const sorted = [...this.image.matchedTags].sort(sortByGenerality);
    return sorted[0];
  }
}

function sortByGenerality(a: ImageTag, b: ImageTag) {
  switch (true) {
    case !!a.version && !!b.version:
      return countOfDots(a.version!) - countOfDots(b.version!);
    case !a.version && !b.version:
      return 0;
    case !a.version:
      return -1;
    case !b.version:
      return 1;
  }
  throw new Error('unreachable');
}

function countOfDots(input: string) {
  return [...input].filter(c => c === '.').length;
}
