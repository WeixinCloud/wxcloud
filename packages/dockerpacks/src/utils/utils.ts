export function optArg(flag: boolean, value: string | string[]) {
  return flag ? (Array.isArray(value) ? value : [value]) : [];
}
