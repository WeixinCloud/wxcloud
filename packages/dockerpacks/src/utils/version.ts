export const DOCKERPACKS_VERSION = process.env.DOCKERPACKS_VERSION || 'unknown';
