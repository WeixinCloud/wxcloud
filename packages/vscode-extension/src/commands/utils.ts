import * as vscode from 'vscode';
import ext from '../core/global';

interface ICommonCommandHandlingOptions {
  showError?: boolean;
  throwError?: boolean;
  refreshOnFinish?: boolean;
}

export function withCommonCommandHandling(
  fn: (...$args: any[]) => any,
  options: ICommonCommandHandlingOptions = {}
) {
  const { showError = true, throwError = true, refreshOnFinish = true } = options;

  return async (...args: any[]) => {
    try {
      return await fn(...args);
    } catch (e) {
      showError && vscode.window.showErrorMessage(`${e}`);
      if (throwError) {
        throw e;
      }
    } finally {
      refreshOnFinish && ext.wxContainersProvider.refresh();
    }
  };
}
