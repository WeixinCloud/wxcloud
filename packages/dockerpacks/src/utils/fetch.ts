import { merge } from 'lodash';
import fetch, { RequestInfo, RequestInit } from 'node-fetch';
import { DOCKERPACKS_VERSION } from './version';

export function fetchUrl(url: RequestInfo, init?: RequestInit) {
  const source: RequestInit = {
    headers: { 'User-Agent': `Dockerpacks ${DOCKERPACKS_VERSION}` }
  };
  return fetch(url, merge(init ?? {}, source));
}
