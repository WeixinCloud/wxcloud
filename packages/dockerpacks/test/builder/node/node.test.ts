import { describe } from 'vitest';
import { BuilderTestCase, runTest } from '@test/utils';
import path from 'path';

const CASES: BuilderTestCase[] = [
  {
    id: 'build',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'none',
    promptAnswers: {
      generalEntrypoint: 'node test/index.js',
      environments: []
    }
  },
  {
    id: 'package-lock',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'npm',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'yarn1',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'yarn2',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'pnpm',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'create-react-app',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'starters-nuxt',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'starters-expressjs',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'starters-sveltekit',
    promptAnswers: {
      generalEntrypoint: 'node build/index.js',
      environments: []
    }
  },
  {
    id: 'starters-nextjs-prisma',
    promptAnswers: {
      environments: ['DATABASE_URL=mongo://test']
    }
  },
  {
    id: 'sveltekit-cli-node-adapter',
    promptAnswers: {
      environments: [],
      generalEntrypoint: 'node build/index.js'
    }
  },
  {
    id: 'sveltekit-cli-static-adapter',
    expectPanic: /adapter-static/i,
    promptAnswers: {
      environments: [],
      generalEntrypoint: 'node build/index.js'
    }
  },
  {
    id: 'nextjs-cli',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'nextjs-cli-pnpm',
    promptAnswers: {
      environments: []
    }
  }
];

const FIXTURES_PATH = path.join(__dirname, 'fixtures');

describe('builders test (node cases)', () => {
  describe.each(CASES)('case %#', testCase => {
    runTest(FIXTURES_PATH, testCase);
  });
});
