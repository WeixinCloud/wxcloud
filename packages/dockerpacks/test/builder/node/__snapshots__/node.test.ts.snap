// Vitest Snapshot v1

exports[`builders test (node cases) > case 0 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

ENV NODE_ENV=production

# 运行项目
CMD npm run start:prod",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 1 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

ENV NODE_ENV=production

# 运行项目
CMD node test/index.js",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md",
  },
}
`;

exports[`builders test (node cases) > case 2 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 安装指定版本的 npm
RUN npm install -g npm@5.10.0

# 设置容器内的当前目录
WORKDIR /app

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

ENV NODE_ENV=production

# 运行项目
CMD node index.js",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 3 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

ENV NODE_ENV=production

# 运行项目
CMD node index.js",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 4 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 安装 yarn
RUN npm install -g yarn@2.4.3

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --immutable

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

ENV NODE_ENV=production

# 运行项目
CMD node index.js",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 5 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN yarn run build

ENV NODE_ENV=production

# 运行项目
CMD npm run start",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 6 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN yarn run build

ENV NODE_ENV=production

# 运行项目
CMD npm run start",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 7 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN yarn run build

ENV NODE_ENV=production

# 运行项目
CMD node build/index.js",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 8 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 用户设置的环境变量
ENV DATABASE_URL=mongo://test

# 将这些文件拷贝到容器中
COPY package.json yarn.lock ./

# 安装依赖
RUN yarn install --non-interactive --prefer-offline

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN yarn run build

ENV NODE_ENV=production

# 运行项目
CMD npm run start",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;

exports[`builders test (node cases) > case 9 > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用基于 alpine 的 node 官方镜像
FROM node:lts-alpine

# 设置容器内的当前目录
WORKDIR /app

# 将 package.json 拷贝到容器中
COPY package.json .

# 安装依赖
RUN npm ci

# 将包括源文件在内的所有文件拷贝到容器中（在 .dockerignore 中的文件除外）
COPY . .

# 运行编译
RUN npm run build

ENV NODE_ENV=production

# 运行项目
CMD npm run start",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md
node_modules",
  },
}
`;
